src/PowerShellEditorServices.Hosting/HostLogger.cs

```diff
@@ -0,0 +1,362 @@
+                    _ui.WriteErrorLine(value.message);
+                    return;
+            }
+        }
+    }
+
+    internal class StreamLogger : IObserver<(PsesLogLevel logLevel, string message)>, IDisposable
+    {
+        public static StreamLogger CreateWithNewFile(string path)
+        {
+            var fileStream = new FileStream(
+                path,
+                FileMode.Create,
+                FileAccess.Write,
+                FileShare.Read,
+                bufferSize: 4096,
+                FileOptions.SequentialScan);
+
+            return new StreamLogger(new StreamWriter(fileStream, new UTF8Encoding(encoderShouldEmitUTF8Identifier: true)));
+        }
+
+        private readonly StreamWriter _fileWriter;
+
+        private readonly BlockingCollection<string> _messageQueue;
+
+        private readonly CancellationTokenSource _cancellationSource;
+
+        private readonly Task _writerTask;
+
+        // This cannot be a bool
+        // See https://stackoverflow.com/q/6164751
+        private int _hasCompleted;
+
+        private IDisposable _unsubscriber;
+
+        public StreamLogger(StreamWriter streamWriter)
+        {
+            streamWriter.AutoFlush = true;
+            _fileWriter = streamWriter;
+            _hasCompleted = 0;
+            _cancellationSource = new CancellationTokenSource();
+            _messageQueue = new BlockingCollection<string>();
+
+            // Start writer listening to queue
+            Task.Run(RunWriter);
+        }
+
+        public void OnCompleted()
+        {
+            // Ensure we only complete once
+            if (Interlocked.Exchange(ref _hasCompleted, 1) != 0)
+            {
+                return;
+            }
+
+            _cancellationSource.Cancel();
+
+            _writerTask.Wait();
+
+            _unsubscriber.Dispose();
+            _fileWriter.Flush();
+            _fileWriter.Close();
+            _fileWriter.Dispose();
+        }
+
+        public void OnError(Exception error)
+        {
+            OnNext((PsesLogLevel.Error, $"Error occurred while logging: {error}"));
+        }
+
+        public void OnNext((PsesLogLevel logLevel, string message) value)
+        {
+            string message = null;
+            switch (value.logLevel)
+            {
+                case PsesLogLevel.Diagnostic:
+                    message = $"[DEBUG]: {value.message}";
+                    break;
+
+                case PsesLogLevel.Verbose:
+                    message = $"[VERBOSE]: {value.message}";
+                    break;
+
+                case PsesLogLevel.Normal:
+                    message = $"[INFO]: {value.message}";
+                    break;
+
+                case PsesLogLevel.Warning:
+                    message = $"[WARN]: {value.message}";
+                    break;
+
+                case PsesLogLevel.Error:
+                    message = $"[ERROR]: {value.message}";
+                    break;
+            };
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this empty statement.](https://app.codacy.com/app/TylerLeonhardt/PowerShellEditorServices/pullRequest?prid=4579423)